### ISSUE TITLE
Some questions about the model dataset

### ISSUE THREAD

**Issue by zuowanbushiwo** [RicherMans/SAT#6]
Hi 
Thank you for your awesome work and sharing it! 

1.   The Training Pipeline mentioned in the paper has three steps, why is there only two steps here?  

> Second, we finetune the pretrained transformer model on Audioset using full-context training (10s).  

The training script for this step was not found. Why?
      
2.   Does the entire Training Pipeline process only use the AudioSet balanced dataset?

In your article 4.1. Datasets, you mentioned that 1,904,746 training samples (5200 h) were used.   But in your `1_run_mae.py` and `2_train_sat.py` ，the `train_data `in config file only has audioset balanced data ， why

Thanks
Best，

**Comment by RicherMans** [comment]
Hey,
thanks for the interest!

> The Training Pipeline mentioned in the paper has three steps, why is there only two steps here?

I didn't think it was necessary for all three steps, i.e., the pretraining of the teacher model can be done with any of my other code bases like [this one](https://github.com/RicherMans/CED).

> Does the entire Training Pipeline process only use the AudioSet balanced dataset?

The full training pipeline used the full audioset dataset, of which the balanced dataset is a subset.

> In your article 4.1. Datasets, you mentioned that 1,904,746 training samples (5200 h) were used. But in your 1_run_mae.py and 2_train_sat.py ，the train_data in config file only has audioset balanced data ， why

Its pretty easy to tell why: For somebody to just "verify" the code, you would need to download the full audioset and downloading audioset is for most people too cumbersome. 
Downloading the balanced subset however, using my code, should take less than couple of hours.
Lastly, the balanced dataset is good enough for a rough pretraining in most cases.

Kind regards,
Heinrich

**Comment by zuowanbushiwo** [comment]
Hi  Heinrich

Thank you for your response, which has helped me understand many issues. As you mentioned, "Just the record, SAT is mainly focused on real-world applications and not tailored towards 'best performance on dataset x'." Therefore, I am very interested in SAT because I currently want to use it in actual products. 

At the same time, I find SAT somewhat complex because it is related to other research projects of yours.

1. PSL - PSEUDO STRONG LABELS FOR LARGE SCALE WEAKLY SUPERVISED AUDIO TAGGING
2. AST - Streaming Audio Transformers for Online Audio Tagging
3. CED - CONSISTENT ENSEMBLE DISTILLATION FOR AUDIO TAGGING

According to the timeline, you published PSL first, followed by AST, then CED.

In AST's Table 1 (Ours (ViT-T), Ours (ViT-S), Ours (ViT-B)) and CED's Table 4 (CED-Tiny, CED-Small, CED-Base),
<img width="406" alt="image" src="https://github.com/user-attachments/assets/1d32feac-7565-48ac-94e5-2325d1560d6d" />

<img width="372" alt="image" src="https://github.com/user-attachments/assets/930ef07f-253d-4855-b0a8-fbe21de603cb" />


1 . Are they the same models?
Because their model sizes are the same, and you mentioned above that "the pretraining of the teacher model can be done with any of my other code bases," so they should be the same, right?

2 . Were they all trained and tested using the 10s data from AudioSet-2M, with mAP tested on AudioSet-eval 10s?

3. What is the reason for the difference in mAP scores (44.2->48.1, 45.7->49.6, 47.4->50.0)? Is it because the training methods were different?
AST used MAE Self-supervised pretraining first, then fine-tuned the pretrained transformer model on Audioset using full-context training, while CED used more advanced ENSEMBLE DISTILLATION?

I have a few points that I don't understand:

1.  In your SAT project, I don't see the influence of the MAE process on SAT training at all, but in Figure 1 of the SAT paper, it shows that the results after MAE training are supposed to initialize SAT training?

2 . How should I train sat_t_1s with your new CED models ? When I try to train sat_t_1s, do I need psl_labels like balanced_vit_base_chunked2s_topk10_47_40.parquet? Is this generated using the PSL project? However, that project only uses the MobileNetV2_DM model, not find  AST, ViT, or CED?

3 . What do you mean by "If that would be the case I recommend you to use the original labels + the PSL labels generated from one of my models as training targets"  [here ](https://github.com/RicherMans/SAT/issues/4#issuecomment-2010879777)? Is there an implementation of this on the project?  I also think this is a good approach., but how do you combine the original labels and PSL labels into one training target?

Thanks！

**Comment by RicherMans** [comment]
Hey,
 > PSL - PSEUDO STRONG LABELS FOR LARGE SCALE WEAKLY SUPERVISED AUDIO TAGGING
> AST - Streaming Audio Transformers for Online Audio Tagging
> CED - CONSISTENT ENSEMBLE DISTILLATION FOR AUDIO TAGGING

Yeah that timelime is correct! Thanks for pointing that out. 

> Because their model sizes are the same, and you mentioned above that "the pretraining of the teacher model can be done with any of my other code bases," so they should be the same, right?

The teachers are just my old "custom" pre-CED trained models on Audioset, so no, its not CED. Before I just recommended to nowadays use CED if you can as a teacher, should work a bit better. 

> Were they all trained and tested using the 10s data from AudioSet-2M, with mAP tested on AudioSet-eval 10s?

Yes. Only in SAT I used 1/2s crops to show how well these models perform with "short delay", i.e., you just need to input 1s of audio instead of waiting for 10s.

> What is the reason for the difference in mAP scores (44.2->48.1, 45.7->49.6, 47.4->50.0)? Is it because the training methods were different?

Yeah, the training methods were different. 
The results in SAT used the usual "hard" Audioset labels, while the results in CED completely drop all original labels and only rely on the "estimated" teacher labels.
CED is pretty good I guess, as seen in this table:

![Image](https://github.com/user-attachments/assets/73bd74ab-4b96-4ab1-bde8-059a92f511ef)

Just by appliying the exact same augmentations between teacher and student training, one can achieve ~+5-7%  absolute in mAP on the balanced subset of Audioset. 

> AST used MAE Self-supervised pretraining first, then fine-tuned the pretrained transformer model on Audioset using full-context training, while CED used more advanced ENSEMBLE DISTILLATION?

Right, that sums it up perfectly, so you can just replace the teacher training with any CED model that is available.


> In your SAT project, I don't see the influence of the MAE process on SAT training at all, but in Figure 1 of the SAT paper, it shows that the results after MAE training are supposed to initialize SAT training?

Well, these vision-transformer based approaches really don't work well without initialization, they just lack some well-thought-out position bias. Most methods that work without pretraining generally rely on some sort of convolutional bias to make training work. Without pretraining, performance usually declines by at least 2 points mAP absolute, but can be more.
However MAE pretraining is pretty fast, so there is literally no downside to it, even on a single GPU, should take not longer than a day.

>  How should I train sat_t_1s with your new CED models ? When I try to train sat_t_1s, do I need psl_labels like balanced_vit_base_chunked2s_topk10_47_40.parquet? Is this generated using the PSL project? However, that project only uses the MobileNetV2_DM model, not find AST, ViT, or CED?

Yeah so that is a bit tricky here. You don't need these .parquet files actually. I implemented most of the logic already in CED. It should be [here](https://github.com/RicherMans/CED/blob/60a0f54214ea619265e16ee257925158be2675fd/utils.py#L26). I set a chunk_length parameter when prediting labels. So you can just set it to like 1/2/3s and estimate novel labels on your dataset. All in all, sorry but I generally like to have my projects containing "the bare minimum", i.e., I provide you with the necessary labels, but the other stuff you need to do yourself. Actually the main problem here is once again that audioset is not an "open source" dataset, thus nearly everybody that downloads has: 1. Different names for the audio (depends on how you save it), 2. different data overall, due to deletion etc.

>  I also think this is a good approach., but how do you combine the original labels and PSL labels into one training target?

So the idea is just to do standard teacher-student training, where you have both "hard" (ground truth) and "soft" (model predictions) targets. You just compute two losses, one for each label. 
The two target approach generally works, but the main problem that we faced for SAT is the lack of "hard" targets for 500ms/1s/2s/3s etc. audio tagging tasks. Strong labels do not solve the issue, since they are costly and data is scarce.

So SAT is a possible approach on howto generate a model with any possible time resolution without needing manual labor or precise labels.

Kind regards,
Heinrich

**Comment by zuowanbushiwo** [comment]
Hi  Heinrich

Thank you very much for your detailed reply; they have helped me clarify many issues. 
first ,I will spend more time reviewing the papers and open-source code related to PSL, SAT, and CED,   Then engaging in more practical operations, and then train my own models.

Best，

**Comment by RicherMans** [comment]
Hey @zuowanbushiwo ,
again thanks for the interest!

If I may add, all three papers (PSL,SAT,CED) are all deployed in our own products (Car, Phone, Speaker, AC and so on).
You might notice that they actually just do the same thing over and over: using teacher labels to predict (soft) student targets.

I would summarize the papers the following:

1. PSL was necessary to deploy models / pretrain models to use any chunk size (500ms/1s, etc), since again in real-world applications nobody waits for 10s and the trained model's performance on its own target chunk-size is usually much better ( usually at least 20 points mAP on real world datasets ) than the 10s alternative.
2. While PSL offers great performance, we wanted to explore streaming models when deploying on stationary devices ( as the paper suggests). As the paper further states, SAT's goal was to be used as a abnormal detector for long-running events, like water running, sound of gas, fire.
3. Lastly, CED is simply PSL with augmentation. Main motivation here was the slow training speed of PSL, since it was only online teacher-student training, which lead to the exclusive usage of small teacher models (Otherwise would be even slower). A reader might think that CED is related to [Efficient-AT](https://arxiv.org/abs/2211.04772), but in reality it had nothing to do with it, i.e., we solved a very different problem. After implementing CED, training speeds greatly increased as did performance.

Hope it helped.
Regards,
Heinrich
